enterprise/app/root/root.tsx

```diff
@@ -83,10 +81,8 @@ export default class EnterpriseRootComponent extends React.Component {
     capabilities.didNavigateToPath();
   }
 
-  handleToggleDenseClicked() {
-    let newDenseMode = !this.state.denseMode;
-    this.setState({ ...this.state, denseMode: newDenseMode });
-    window.localStorage.setItem(denseModeKey, newDenseMode ? denseModeValue : "");
+  handlePreferencesChanged() {
+    this.setState({ ...this.state, preferences: this.state.preferences });
```

Comment:
I think you need to clone the object, otherwise child components will see that the prop didn't change, and won't re-render

Reply:
Might want to make UserPreferences a global like we do with rpcService etc. then expose an observable preferences object which is just a plain old JS object -- that way you can listen to changes using the `subscribe()` API and then clone it easily with `this.setState({ preferences: {...preferences.values}})`

Reply:
According to [the docs](https://reactjs.org/docs/react-component.html#setstate), setState will always re-render unless you implement shouldComponentUpdate which we don't anywhere.

> setState() will always lead to a re-render unless shouldComponentUpdate() returns false. If mutable objects are being used and conditional rendering logic cannot be implemented in shouldComponentUpdate(), calling setState() only when the new state differs from the previous state will avoid unnecessary re-renders.

Reply:
Ah, I'm thinking of `React.PureComponent` which we aren't using anywhere yet. Pure components don't re-render if their props don't change -- https://codesandbox.io/s/interesting-nobel-nmkgu?file=/src/App.js

Pure components are an easy way to improve performance if re-render time becomes a bottleneck, and it would be great if we could avoid introducing patterns that break from the "reactive" model and make it harder to use pure components down the road (like classes whose reference doesn't change, but their internal state does). As of now I think everything could be made pure, except components which use InvocationModel.

Reply:
I think components that have 'User' as a prop (and their children which is pretty much every component) would also be an issue if we wanted to use PureComponents.

There are also quite a few components that have mutable objects as part of their state. 

It's a nice goal, but we haven't focused on immutability in state / props. If render performance becomes an issue, we're always going to have to go component by component and enable that optimizations for components that support it - and make things immutable before converting components that don't.